### PR TITLE
Fix 13509

### LINKED ISSUE #13509: Instantiating an exception with wrong payload compiles and fails at run-time
It is possible to instantiate a user defined exception by giving an interpolated string (without anything interpolated) as a constructor argument without compile-time error although the constructor expects something else.

Note that the problem does not reproduce (i.e. compile error is given as expected) if normal string is used instead of interpolated string or something is interpolated in the interpolated string.

**Repro steps**

1. Create new console application
2. Add the code below to Program.fs
3. Run `dotnet run`

```fsharp
type Foo = { Bar: int }

exception MyException of string * Foo

try
    raise (MyException($"BOOM"))
    // raise (MyException("BOOM")) // <- Compile-time error
    // raise (MyException($"BOOM {42}")) // <- Compile-time error
with
| MyException(msg, foo) ->
    printfn "%A" foo
```

**Expected behavior**

Code should not compile.

**Actual behavior**

Code compiles, but fails at run-time with following error:
```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.Runtime.CompilerServices.CastHelpers.ChkCastAny(Void*, System.Object)
   at Microsoft.FSharp.Core.LanguagePrimitives+IntrinsicFunctions.UnboxGeneric[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]](System.Object)
   at Microsoft.FSharp.Core.PrintfImpl+GenericToString@969-3[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Invoke(System.Object)
   at Microsoft.FSharp.Core.PrintfImpl+OneStepWithArg@501-1[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].Invoke(System.__Canon)
   at <StartupCode$ExceptionErrorReproduction>.$Program.main@()
```

**Related information**

* Operating system: macOS 12.3.1
* .NET SDK version: 6.0.302

### PR DESCRIPTION
**Pull request by vzarytovskii**
Fixes #13509
This fixes binding, need to check exceptions.

### CODE REVIEW

**Review by KevinRansom — APPROVED** [review]

**Comment by vzarytovskii** [comment]
Fixes exceptions too

**Comment by vzarytovskii** [comment]
This is ready, do we need more tests added?

File: tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    [<Fact>]
+    let ``Regression: Empty Interpolated String properly typechecks with explicit type on binding`` () =
+        Fsx """ let a:byte = $"string" """
+        |> compile
+        |> shouldFail
+        |> withSingleDiagnostic (Error 1, Line 1, Col 15, Line 1, Col 24, "This expression was expected to have type" + Environment.NewLine + "    'byte'    " + Environment.NewLine + "but here has type" + Environment.NewLine + "    'string'    ")
+
+    [<Fact>]
+    let ``Interpolated String with hole properly typechecks with explicit type on binding`` () =
+        Fsx """ let a:byte = $"string" """
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs:20]: You probably forgot to put the hole there.

**Review by 0101 — APPROVED** [review]

File: tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    [<Fact>]
+    let ``Regression: Empty Interpolated String properly typechecks with explicit type on binding`` () =
+        Fsx """ let a:byte = $"string" """
+        |> compile
+        |> shouldFail
+        |> withSingleDiagnostic (Error 1, Line 1, Col 15, Line 1, Col 24, "This expression was expected to have type" + Environment.NewLine + "    'byte'    " + Environment.NewLine + "but here has type" + Environment.NewLine + "    'string'    ")
+
+    [<Fact>]
+    let ``Interpolated String with hole properly typechecks with explicit type on binding`` () =
```
Comment (dsyme) [inline tests/FSharp.Compiler.ComponentTests/Language/InterpolatedStringsTests.fs:19]: Could you add tests for

    let a: obj = $"string"
    let a: System.IComparable = $"string"
    let a: System.IFormattable = $"string"

**Review by dsyme — APPROVED** [review]

**Review by dsyme — APPROVED** [review]

**Pull request merged**